build/helper/metadata_add_all.py

```diff
@@ -440,57 +440,103 @@ def add_all_attribute_metadata(attributes, config):
         for p in config['functions'][f]['parameters']:
             e = p['enum']
-            if e is not None and e not in enums:
-                print('Missing enum {0} referenced by function {1}'.format(e, f))
-            elif e is not None:
-                if f_codegen_method == 'private' and enums[e]['codegen_method'] == 'no':
-                    enums[e]['codegen_method'] = f_codegen_method
-                elif f_codegen_method in ('public', 'python-only') and enums[e]['codegen_method'] != 'public':
-                    if enum_has_explicit_codegen_method[e]:
-                        if not enum_uses_converter(enums[e]):
-                            raise ValueError(
-                                f"Enum {e} used by public function {f} must be public, is private"
-                            )
-                    else:
-                        enums[e]['codegen_method'] = 'public'
-
-    # Iterate through all codegen attributes and set any enum parameters to the same level of codegen_method
+            if e is not None:
+                if e not in enum_to_client_functions:
+                    print('Missing enum {0} referenced by function {1}'.format(e, f))
+                else:
+                    enum_to_client_functions[e].append(f)
+    return enum_to_client_functions
+
+
+def _get_attributes_that_use_enums(enums, config):
+    '''Generate a dict that maps each enum to a list of attributes that use it'''
+    enum_to_client_attributes = {e: [] for e in enums}
     for a in filter_codegen_attributes(config['attributes']):
-        a_codegen_method = config['attributes'][a]['codegen_method']
         e = config['attributes'][a]['enum']
-        if e is not None and e not in enums:
-            print('Missing enum {0} referenced by attribute {1}'.format(e, a))
-        elif e is not None:
-            if a_codegen_method == 'private' and enums[e]['codegen_method'] == 'no':
-                enums[e]['codegen_method'] = a_codegen_method
-            elif a_codegen_method == 'public' and enums[e]['codegen_method'] != 'public':
-                if enum_has_explicit_codegen_method[e]:
-                    if not enum_uses_converter(enums[e]):
-                        raise ValueError(
-                            f"Enum {e} used by public attribute {config['attributes'][a]['name']} must be public, is private"
-                        )
-                else:
-                    enums[e]['codegen_method'] = a_codegen_method
+        if e is not None:
+            if e not in enum_to_client_attributes:
+                print('Missing enum {0} referenced by attribute {1}'.format(e, a))
+            else:
+                enum_to_client_attributes[e].append(a)
+    return enum_to_client_attributes
+
+
+def _get_least_restrictive_codegen_method(codegen_methods):
+    '''Get the least restrictive codegen_method among the input codegen_methods list.
+
+    If the input codegen_methods list, return 'no'.
```

Comment:
```suggestion
    If codegen_methods parameter is empty, return 'no'.
```

Reply:
Thanks for spotting this, I have fixed it in my latest commit.